src/main.rs

```diff
@@ -831,145 +491,9 @@ impl App {
-                        format!("Failed to create recording pipeline: {}", err).as_str(),
-                    );
-                    return;
-                }
-                Ok(bin) => bin,
-            };
-
-            // Get our file sink element by its name and set the location where to write the recording
-            let sink = bin
-                .get_by_name("sink")
-                .expect("Recording bin has no sink element");
-            let mut filename = settings.record_directory.clone();
-            let now = Local::now();
-            filename.push(format!(
-                "{}.{}",
-                now.format("Recording %Y-%m-%d %H:%M:%S"),
-                extension
-            ));
-
-            // All strings in GStreamer are UTF8, we need to convert the path to UTF8
-            // which in theory can fail
-            sink.set_property("location", &(filename.to_str().unwrap()))
-                .expect("Filesink had no location property");
-
-            // First try setting the recording bin to playing: if this fails
-            // we know this before it potentially interferred with the other
-            // part of the pipeline
-            if let Err(_) = bin.set_state(gst::State::Playing).into_result() {
-                show_error_dialog(
-                    self.0.borrow().main_window.as_ref(),
-                    false,
-                    "Failed to start recording",
-                );
-                return;
-            }
-
-            // Add the bin to the pipeline. This would only fail if there was already
-            // a bin with the same name, which we ensured can't happen
-            pipeline.add(&bin).expect("Failed to add recording bin");
-
-            // Get our tee element by name, request a new source pad from it and
-            // then link that to our recording bin to actually start receiving data
-            let tee = pipeline
-                .get_by_name("tee")
-                .expect("Pipeline had no tee element");
-            let srcpad = tee
-                .get_request_pad("src_%u")
-                .expect("Failed to request new pad from tee");
-            let sinkpad = bin
-                .get_static_pad("sink")
-                .expect("Failed to get sink pad from recording bin");
-
-            // If linking fails, we just undo what we did above
-            if let Err(err) = srcpad.link(&sinkpad).into_result() {
-                show_error_dialog(
-                    self.0.borrow().main_window.as_ref(),
-                    false,
-                    format!("Failed to link recording bin: {}", err).as_str(),
-                );
-                // This might fail but we don't care anymore: we're in an error path
-                let _ = pipeline.remove(&bin);
-                let _ = bin.set_state(gst::State::Null);
-            }
-
-            println!("Recording to {}", filename.display());
+            self.start_recording(&pipeline, record_button, settings);
```

Comment:
The problem with moving this is that we now have a footgun: the refcell is borrowed mutable here, but you wouldn't necessarily know that inside the new function.

Maybe the new function should be defined on `AppInner` instead, maybe most of these functions here should instead be defined on `AppInner` instead of `App` so that borrowing the refcell happens exactly once at a defined place? (Problem then: how to get from `AppInner` to a weak ref for `App` from inside those functions)